File: src/backend/screencast_portal.py
# SPDX-FileCopyrightText: Copyright 2018-2021 Jonas Adahl
# SPDX-FileCopyrightText: Copyright 2021 SeaDve
# SPDX-License-Identifier: GPL-3.0-or-later

from enum import IntEnum
from collections import namedtuple

from gi.repository import GObject, GLib, Gio

from kooha.logger import Logger

Screen = namedtuple('Screen', 'w h')
Stream = namedtuple('Stream', 'fd node_id screen')


class Response(IntEnum):
    SUCCESS = 0
    CANCELLED = 1
    FAILED = 2


class ScreencastPortal(GObject.GObject):
    """Opens and closes pipewire stream to be used by the pipeline in Recorder"""

    __gsignals__ = {'ready': (GObject.SignalFlags.RUN_FIRST, None, (object, bool)),
                    'revoked': (GObject.SignalFlags.RUN_FIRST, None, (str,))}

    def __init__(self):
        super().__init__()

        self.bus = Gio.bus_get_sync(Gio.BusType.SESSION, None)
        self.proxy = Gio.DBusProxy.new_sync(
            self.bus,
            Gio.DBusProxyFlags.NONE,
            None,
            'org.freedesktop.portal.Desktop',
            '/org/freedesktop/portal/desktop',
            'org.freedesktop.portal.ScreenCast',
            None
        )

        self.sender_name = self.bus.get_unique_name()[1:].replace('.', '_')
        self.request_counter = 0
        self.session_counter = 0

    def _on_create_session_response(self, bus, sender, path, request_path, node, output):
        response, results = output
        if response != Response.SUCCESS:
            self.emit('revoked', _("Failed to create session."))
            Logger.warning(f"Failed to create session: {response}")
            return

        self.session_handle = results['session_handle']
        Logger.info("Session created")
        self._screencast_call(
            self.proxy.SelectSources,
            self._on_select_sources_response,
            '(oa{sv})',
            self.session_handle,
            options={
                'types': GLib.Variant('u', 1 if self.is_selection_mode else 1 | 2),
                'cursor_mode': GLib.Variant('u', 2 if self.is_show_pointer else 1)
            }
        )

    def _on_select_sources_response(self, bus, sender, path, request_path, node, output):
        response, results = output
        if response != Response.SUCCESS:
            self.emit('revoked', _("Failed to select sources."))
            Logger.warning(f"Failed to select sources: {response}")
            return

        Logger.info("Sources selected")
        self._screencast_call(
            self.proxy.Start,
            self._on_start_response,
            '(osa{sv})',
            self.session_handle,
            ''
        )

    def _on_start_response(self, bus, sender, path, request_path, node, output):
        response, results = output
        if response == Response.CANCELLED:
            self.emit('revoked', None)
            Logger.info("Interaction cancelled")
            return

        if response == Response.FAILED:
            self.emit('revoked', _("Failed to start."))
            Logger.warning(f"Failed to start: {response}")
            return

        fd = self._get_fd()
        node_id, stream_info = results['streams'][0]
        stream_screen = Screen(*stream_info['size'])
        pipewire_stream = Stream(fd, node_id, stream_screen)

        self.emit('ready', pipewire_stream, self.is_selection_mode)
        Logger.info("Ready for pipewire stream")

    def _new_session_path(self):
        self.session_counter += 1
        token = f'u{self.session_counter}'
        path = f'/org/freedesktop/portal/desktop/session/{self.sender_name}/{token}'
        return path, token

    def _new_request_path(self):
        self.request_counter += 1
        token = f'u{self.request_counter}'
        path = f'/org/freedesktop/portal/desktop/request/{self.sender_name}/{token}'
        return path, token

    def _get_fd(self):
        response, fd_list = self.proxy.call_with_unix_fd_list_sync(
            'OpenPipeWireRemote',
            GLib.Variant('(oa{sv})', (self.session_handle, {})),
            Gio.DBusCallFlags.NONE,
            -1,
            None,
            None
        )
        return fd_list.get(0)

    def _screencast_call(self, method, callback, signature, *args, options={}):
        request_path, request_token = self._new_request_path()
        self.bus.signal_subscribe(
            'org.freedesktop.portal.Desktop',
            'org.freedesktop.portal.Request',
            'Response',
            request_path,
            None,
            Gio.DBusSignalFlags.NONE,
            callback
        )
        options['handle_token'] = GLib.Variant('s', request_token)

        try:
            method(signature, *args, options)
        except GLib.Error as error:
            self.emit('revoked', error)
            Logger.warning(error)

    def open(self, is_show_pointer, is_selection_mode):
        self.is_show_pointer = is_show_pointer
        self.is_selection_mode = is_selection_mode

        _, session_token = self._new_session_path()
        self._screencast_call(
            self.proxy.CreateSession,
            self._on_create_session_response,
            '(a{sv})',
            options={
                'session_handle_token': GLib.Variant('s', session_token),
            }
        )

    def close(self):
        session_proxy = Gio.DBusProxy.new_sync(
            self.bus,
            Gio.DBusProxyFlags.NONE,
            None,
            'org.freedesktop.portal.Desktop',
            self.session_handle,
            'org.freedesktop.portal.Session',
            None
        )
        session_proxy.Close()
        Logger.info("Portal closed")

File: src/backend/screencast_portal.rs
use ashpd::{
    desktop::{
        screencast::{CursorMode, PersistMode, ScreenCastProxy, SourceType, Stream},
        ResponseError, SessionProxy,
    },
    enumflags2::BitFlags,
    zbus, WindowIdentifier,
};
use gtk::{
    glib::{self, clone},
    subclass::prelude::*,
};

use std::{cell::RefCell, os::unix::io::RawFd};

use crate::{error::Error, settings::Settings, Application};

#[derive(Debug)]
pub enum ScreencastPortalResponse {
    Success(Vec<Stream>, i32),
    Failed(Error),
    Cancelled,
}

mod imp {
    use super::*;

    #[derive(Debug, Default)]
    pub struct ScreencastPortal {
        pub session: RefCell<Option<SessionProxy<'static>>>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for ScreencastPortal {
        const NAME: &'static str = "KoohaScreencastPortal";
        type Type = super::ScreencastPortal;
    }

    impl ObjectImpl for ScreencastPortal {}
}

glib::wrapper! {
    pub struct ScreencastPortal(ObjectSubclass<imp::ScreencastPortal>);
}

impl ScreencastPortal {
    pub fn new() -> Self {
        glib::Object::new::<Self>(&[]).expect("Failed to create ScreencastPortal.")
    }

    fn cursor_mode(is_show_pointer: bool) -> BitFlags<CursorMode> {
        if is_show_pointer {
            BitFlags::<CursorMode>::from_flag(CursorMode::Embedded)
        } else {
            BitFlags::<CursorMode>::from_flag(CursorMode::Hidden)
        }
    }

    fn source_type(is_selection_mode: bool) -> BitFlags<SourceType> {
        if is_selection_mode {
            BitFlags::<SourceType>::from_flag(SourceType::Monitor)
        } else {
            SourceType::Monitor | SourceType::Window
        }
    }

    fn settings() -> Settings {
        Application::default().settings()
    }

    async fn identifier() -> WindowIdentifier {
        let main_window = Application::default().main_window();
        WindowIdentifier::from_native(&main_window).await
    }

    pub async fn new_session(
        &self,
        is_show_pointer: bool,
        is_selection_mode: bool,
    ) -> ScreencastPortalResponse {
        let identifier = Self::identifier().await;
        let multiple = !is_selection_mode;
        let source_type = Self::source_type(is_selection_mode);
        let cursor_mode = Self::cursor_mode(is_show_pointer);
        let restore_token = Self::settings().screencast_restore_token();

        match screencast(
            identifier,
            multiple,
            source_type,
            cursor_mode,
            restore_token.as_deref(),
        )
        .await
        {
            Ok(result) => {
                let (streams, fd, restore_token, session) = result;
                self.imp().session.replace(Some(session));
                Self::settings().set_screencast_restore_token(restore_token.as_deref());

                ScreencastPortalResponse::Success(streams, fd)
            }
            Err(error) => match error {
                ashpd::Error::Response(ResponseError::Cancelled) => {
                    log::info!("Select sources cancelled");
                    ScreencastPortalResponse::Cancelled
                }
                other_error => {
                    log::error!("Error from screencast call: {:?}", other_error);
                    ScreencastPortalResponse::Failed(Error::from(other_error))
                }
            },
        }
    }

    pub fn close_session(&self) {
        let ctx = glib::MainContext::default();
        ctx.spawn_local(clone!(@weak self as obj => async move {
            if let Some(session) = obj.imp().session.take() {
                session.close().await.unwrap();
            };
        }));

        log::info!("Session closed");
    }
}

async fn screencast(
    window_identifier: WindowIdentifier,
    multiple: bool,
    types: BitFlags<SourceType>,
    cursor_mode: BitFlags<CursorMode>,
    restore_token: Option<&str>,
) -> Result<(Vec<Stream>, RawFd, Option<String>, SessionProxy<'static>), ashpd::Error> {
    let connection = zbus::Connection::session().await?;
    let proxy = ScreenCastProxy::new(&connection).await?;
    log::info!("ScreenCastProxy created");

    log::debug!("restore_token: {:?}", restore_token);

    log::debug!(
        "available_cursor_modes: {:?}",
        proxy.available_cursor_modes().await?
    );
    log::debug!(
        "available_source_types: {:?}",
        proxy.available_source_types().await?
    );

    let session = proxy.create_session().await?;
    log::info!("Session created");

    proxy
        .select_sources(
            &session,
            cursor_mode,
            types,
            multiple,
            restore_token,
            PersistMode::ExplicitlyRevoked,
        )
        .await?;
    log::info!("Select sources window showed");

    let (streams, output_restore_token) = proxy.start(&session, &window_identifier).await?;
    log::info!("Screencast session started");

    log::debug!("output_restore_token: {:?}", output_restore_token);

    let fd = proxy.open_pipe_wire_remote(&session).await?;
    log::info!("Ready for pipewire stream");

    Ok((streams, fd, output_restore_token, session))
}

impl Default for ScreencastPortal {
    fn default() -> Self {
        Self::new()
    }
}
